bridge/gitlab/import.go

```diff
@@ -0,0 +1,351 @@
+		)
+		return err
+
+	case NOTE_DESCRIPTION_CHANGED:
+		issue := gi.iterator.IssueValue()
+
+		// since gitlab doesn't provide the issue history
+		// we should check for "changed the description" notes and compare issue texts
+		// TODO: Check only one time and ignore next 'description change' within one issue
+		if issue.Description != b.Snapshot().Comments[0].Message {
+
+			// comment edition
+			_, err = b.EditCommentRaw(
+				author,
+				note.UpdatedAt.Unix(),
+				target,
+				issue.Description,
+				map[string]string{
+					keyGitlabId:  id,
+					keyGitlabUrl: "",
+				},
+			)
+
+			return err
+
+		}
+
+	case NOTE_COMMENT:
+
+		cleanText, err := text.Cleanup(body)
+		if err != nil {
+			return err
+		}
+
+		// if we didn't import the comment
+		if errResolve == cache.ErrNoMatchingOp {
+
+			// add comment operation
+			_, err = b.AddCommentRaw(
+				author,
+				note.CreatedAt.Unix(),
+				cleanText,
+				nil,
+				map[string]string{
+					keyGitlabId:  id,
+					keyGitlabUrl: "",
+				},
+			)
+
+			return err
+		}
+
+		// if comment was already exported
+
+		// if note wasn't updated
+		if note.UpdatedAt.Equal(*note.CreatedAt) {
+			return nil
+		}
+
+		// search for last comment update
+		comment, err := b.Snapshot().SearchComment(hash)
+		if err != nil {
+			return err
+		}
+
+		// compare local bug comment with the new note body
+		if comment.Message != cleanText {
+			// comment edition
+			_, err = b.EditCommentRaw(
+				author,
+				note.UpdatedAt.Unix(),
+				target,
+				cleanText,
+				map[string]string{
+					// no metadata unique metadata to store
+					keyGitlabId:  "",
+					keyGitlabUrl: "",
+				},
+			)
+
+			return err
+		}
+
+		return nil
+
+	case NOTE_TITLE_CHANGED:
+		// title change events are given new notes
+		_, err = b.SetTitleRaw(
+			author,
+			note.CreatedAt.Unix(),
+			body,
+			map[string]string{
+				keyGitlabId:  id,
+				keyGitlabUrl: "",
+			},
+		)
+
+		return err
+
+	case NOTE_UNKNOWN:
+		//TODO: send warning via channel
+		return nil
+
+	default:
+		panic("unhandled note type")
+	}
+
+	return nil
+}
+
+func (gi *gitlabImporter) ensureLabelEvent(repo *cache.RepoCache, b *cache.BugCache, labelEvent *gitlab.LabelEvent) error {
+	_, err := b.ResolveOperationWithMetadata(keyGitlabId, parseID(labelEvent.ID))
+	if err != cache.ErrNoMatchingOp {
+		return err
+	}
+
+	// ensure issue author
+	author, err := gi.ensurePerson(repo, labelEvent.User.ID)
+	if err != nil {
+		return err
+	}
+
+	switch labelEvent.Action {
+	case "add":
+		_, err = b.ForceChangeLabelsRaw(
+			author,
+			labelEvent.CreatedAt.Unix(),
+			[]string{labelEvent.Label.Name},
+			nil,
+			map[string]string{
+				keyGitlabId: parseID(labelEvent.ID),
+			},
+		)
+
+	case "remove":
+		_, err = b.ForceChangeLabelsRaw(
+			author,
+			labelEvent.CreatedAt.Unix(),
+			nil,
+			[]string{labelEvent.Label.Name},
+			map[string]string{
+				keyGitlabId: parseID(labelEvent.ID),
+			},
+		)
+
+	default:
+		panic("unexpected label event action")
```

Comment:
panic is harsh, it can happen if gitlab change its API. An error would be better.